collect_app/src/main/java/org/odk/collect/android/tasks/InstanceGoogleSheetsUploader.java

```diff
@@ -68,483 +78,418 @@
-            content.add(list);
-            ValueRange row = new ValueRange();
-            row.setValues(content);
+    private List<Object> getColumnTitles(TreeElement element) {
+        List<Object> columnTitles = new ArrayList<>();
+        if (element.isRepeatable()) {
+            columnTitles.add(0, INSTANCE_ID);
+        }
+        for (TreeElement child : getChildElements(element)) {
+            columnTitles.add(getElementTitle(child));
+        }
+        return columnTitles;
+    }
 
-            try {
-                sheetsHelper.insertRow(spreadsheetId, sheetName + "!A1:1", row);
-            } catch (IOException e) {
-                Timber.e(e);
-                outcome.results.put(id, e.getMessage());
-                return false;
+    private TreeElement getInstanceIDElement(List<TreeElement> elements) {
+        for (TreeElement element : elements) {
+            if (element.getName().equals(INSTANCE_ID)) {
+                return element;
             }
         }
+        return null;
+    }
 
-        // we may have updated the feed, so get a new one
-        // update the feed
-
-        try {
-            values = sheetsHelper.getHeaderFeed(spreadsheetId, sheetName);
-            if (values == null || values.size() == 0) {
-                outcome.results.put(id, "No data found");
-                return false;
-            } else {
-                headerFeed = values.get(0);
+    private boolean doesSheetExist(String sheetTitle) {
+        for (Sheet sheet : spreadsheet.getSheets()) {
+            if (sheet.getProperties().getTitle().equals(sheetTitle)) {
+                return true;
             }
-        } catch (IOException e) {
-            Timber.e(e, "Exception thrown while getting the header feed");
-            outcome.results.put(id, e.getMessage());
-            return false;
         }
+        return false;
+    }
 
-        // first, get all the columns in the spreadsheet
-        ArrayList<String> sheetCols = new ArrayList<>();
-        if (headerFeed != null) {
-            for (Object column : headerFeed) {
-                sheetCols.add(column.toString());
+    private void disallowMissingColumns(List<Object> columnHeaders, List<Object> columnTitles) throws UploadException {
+        for (Object columnTitle : columnTitles) {
+            if (!columnHeaders.contains(columnTitle)) {
+                throw new UploadException(Collect.getInstance().getString(R.string.google_sheets_missing_columns, columnTitle));
             }
-        } else {
-            outcome.results.put(id, "couldn't get header feed");
-            return false;
         }
+    }
 
-        ArrayList<String> missingColumns = new ArrayList<>();
-        for (String col : columnNames) {
-            if (!sheetCols.contains(col)) {
-                missingColumns.add(col);
-            }
+    // This  method builds a column name by joining all of the containing group names using "-" as a separator
```

Comment:
We might as well enjoy the benefits of having this be a doc comment. Pop-up help in the IDE, for instance.
`/** This method builds a column name by joining all of the containing group names using "-" as a separator */
(Please notice the removed extra space.)